allennlp/models/event2mind.py

```diff
@@ -0,0 +1,467 @@
+        # List of (batchsize, k) tensors. One for each time step. None for
+        # the first.  Stores the index n for the parent prediction, i.e.
+        # predictions[t-1][i][n], that it came from.
+        backpointers = []
+
+        # Timestep 1
+        start_predictions = final_encoder_output.new_full(
+                (batch_size,), fill_value=self._start_index, dtype=torch.long
+        )
+        start_decoder_input = target_embedder(start_predictions)
+        start_decoder_hidden = decoder_cell(start_decoder_input, final_encoder_output)
+        start_output_projections = output_projection_layer(start_decoder_hidden)
+        start_class_log_probabilities = F.log_softmax(start_output_projections, dim=-1)
+        start_top_log_probabilities, start_predicted_classes = start_class_log_probabilities.topk(k)
+
+        # Set starting values
+        # The log probabilities for the last time step. (batch_size, k)
+        log_probabilities = start_top_log_probabilities
+        # [(batch_size, k)]
+        predictions.append(start_predicted_classes)
+        # Set the same hidden state for each element in beam.
+        # (batch_size * k, _decoder_output_dim)
+        decoder_hidden = start_decoder_hidden.\
+            unsqueeze(1).expand(batch_size, k, self._decoder_output_dim).\
+            reshape(batch_size * k, self._decoder_output_dim)
+
+        # Log probability tensor that mandates that the end token is selected.
+        num_classes = self.vocab.get_vocab_size(self._target_namespace)
+        log_probs_after_end = start_class_log_probabilities.new_full(
+                (batch_size * k, num_classes),
+                float("-inf")
+        )
+        log_probs_after_end[:, self._end_index] = 0.0
+
+        for timestep in range(num_decoding_steps - 1):
+            # (batch_size * k,)
+            last_predictions = predictions[-1].reshape(batch_size * k)
+            decoder_input = target_embedder(last_predictions)
+            decoder_hidden = decoder_cell(decoder_input, decoder_hidden)
+            # (batch_size * k, num_classes)
+            output_projections = output_projection_layer(decoder_hidden)
+
+            # (batch_size * k, num_classes)
+            class_log_probabilities = F.log_softmax(output_projections, dim=-1)
+
+            # (batch_size * k, num_classes)
+            last_predictions_expanded = last_predictions.unsqueeze(-1).expand(batch_size * k, num_classes)
+            cleaned_log_probabilities = torch.where(
+                    last_predictions_expanded == self._end_index,
+                    log_probs_after_end,
+                    class_log_probabilities
+            )
+
+            # (batch_size * k, k), (batch_size * k, k)
+            top_log_probabilities, predicted_classes = cleaned_log_probabilities.topk(k)
+            # TODO(brendanr): Normalize for length?
+            # (batch_size * k, k)
+            expanded_last_log_probabilities = log_probabilities.\
```

Comment:
Add a comment:

"Here we expand the last log probabilities to (batch_size * beam_size, beam_size) so that we can add them to the current log probs for this timestep. This lets us maintain the log probability of each element on the beam."